### PR TITLE
Add basic support for Gecko atoms

### PR DESCRIPTION
**Pull request by bholley**
This is a rough initial implementation of gecko atoms. This allows us to get rid of the custom rust-selectors branch we use to build stylo.

The gecko changes are at https://bugzilla.mozilla.org/show_bug.cgi?id=1273771

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11242)

### CODE REVIEW

**Comment by nox** [comment]
I yanked selectors 0.5.10. Please amend your bump with selectors 0.6.0.

**Comment by nox** [comment]
Also @SimonSapin is on PTO for the week.

**Comment by nox** [comment]
@bholley Could that BorrowedAtom thing have been made an associated type in a selectors trait? Something like `type BorrowedAtom: Borrow<Atom> = &'a Atom;`?

File: ports/geckolib/string_cache/regen_atom_macro.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+objdir_path = sys.argv[1]
+
+
+def line_to_atom(line):
+    result = re.match('^GK_ATOM\((.+),\s*"(.*)"\)', line)
+    return (result.group(1), result.group(2))
+
+
+def symbolify(ident):
+    return "_ZN9nsGkAtoms" + str(len(ident)) + ident + "E"
```
Comment (emilio) [inline ports/geckolib/string_cache/regen_atom_macro.py]: Bindgen has support for this kind of things, just sayin, and while it cannot generate the macro, it does generate nicer bindings, and would avoid hardcoding the ABI mangling:

[This][bindings] is the result of:

```
./target/debug/bindgen -ignore-functions -no-type-renaming -x c++ \
                       /gecko/objdir/dist/include/nsGkAtoms.h -o nsGkAtoms.rs
```

[bindings]: https://gist.github.com/emilio/d2e1a3c8e5118310b4f61815a16c7f71

**Comment by emilio** [comment]
Most nits are not too relevant, though I'd prefer them to be addressed.

r=me otherwise

-S-awaiting-review

<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> r? @emilio
</details>


---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 22 of 22 files at r3, 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/style/attr.rs, line 300 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3o-25zN569UinRp6j:-KI3o-25zN569UinRp6k:870287907) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/components/style/attr.rs#L300)):*
> ```Rust
> 
> #[cfg(not(feature = "gecko"))] // Gecko can't borrow atoms as UTF-8.
> impl Deref for AttrValue {
> ```

nit: Probably this is cleaner removing the `cfg` at the top and here: `impl ::std::ops::Deref`, leaving just one `#[cfg]`.

---

*[ports/geckolib/string_cache/Cargo.toml, line 1 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3qXdrlCI3ixSVzC_f:-KI3qXdrlCI3ixSVzC_g:1274762074) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/Cargo.toml#L1)):*
> ```TOML
> [package]
> ```

I guess that if we're hosting this in-tree, then https://github.com/servo/gecko-atom should disappear, right?

---

*[ports/geckolib/string_cache/lib.rs, line 118 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3qre1Rg2FbiWpfpOt:-KI3qre1Rg2FbiWpfpOu:10654023) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/lib.rs#L118)):*
> ```Rust
>                                where F: FnOnce(&str) -> Output {
>         // FIXME(bholley): We should measure whether it makes more sense to
>         // cache the UTF-8 version in the Gecko atom table somehow.
> ```

I was going to ask for something like this... Seems like a decent way to implement the utf-8 borrowing without leaks, if it's not too space-expensive... Can't think of a less-expensive way of doing it right now, though I arguably haven't thought too much about it.

---

*[ports/geckolib/string_cache/lib.rs, line 219 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3rSA4VvCkwHdnZ04f:-KI3rSA4VvCkwHdnZ04g:-1606705230) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/lib.rs#L219)):*
> ```Rust
>     #[inline]
>     fn from(string: Cow<'a, str>) -> Atom {
>         Atom::from(&string[..])
> ```

nit: This would (probably?) work  as `&*string`, and we use that a lot more through the rest of the tree.

---

*[ports/geckolib/string_cache/lib.rs, line 226 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3rizUklCEXZImnbln:-KI3rizUklCEXZImnblo:-1119816080) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/lib.rs#L226)):*
> ```Rust
>     #[inline]
>     fn from(string: String) -> Atom {
>         Atom::from(&string[..])
> ```

nit: ditto

---

*[ports/geckolib/string_cache/regen_atom_macro.py, line 19 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI1e4iCNQHlRH_q5YlP-r3-19:-KI3rnIHluhXNr_ex7ow:-2081435563) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/regen_atom_macro.py#L19)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> Bindgen has support for this kind of things, just sayin, and while it cannot generate the macro, it does generate nicer bindings, and would avoid hardcoding the ABI mangling:
> 
> 
> 
> [This][bindings] is the result of:
> 
> 
> 
> ```
> 
> ./target/debug/bindgen -ignore-functions -no-type-renaming -x c++ \
> 
>                        /gecko/objdir/dist/include/nsGkAtoms.h -o nsGkAtoms.rs
> 
> ```
> 
> 
> 
> [bindings]: https://gist.github.com/emilio/d2e1a3c8e5118310b4f61815a16c7f71
</details>
After the IRC talk, I'm fine with landing this as-is, but (for the record) this would probably require at some point either taking in account other manglings like win32's, or using bindgen (that relies in clang for it).

---

*[ports/geckolib/string_cache/regen_atom_macro.py, line 31 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3sAIWr9aTP9p6-pS7:-KI3sAIWr9aTP9p6-pS8:-771751615) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/regen_atom_macro.py#L31)):*
> ```Python
>     f.write("pub fn unsafe_atom_from_static(ptr: *mut nsIAtom) -> Atom { unsafe { Atom::from_static(ptr) } }\n\n")
>     for atom in atoms:
>         f.write('extern { pub static %s: *mut nsIAtom; }\n' % symbolify(atom[0]))
> ```

Writing this in a single `extern "C"` block would make the resulting code a lot more readable. Ideally using something like:

```python
"#[link_name=\"%s\"] pub static %s: *mut nsIAtom;\n" % (symbolify(atom[0]), atom[0])
```

This is generated code though, and it's clear what it does, so I won't insist on this a lot :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11242#-:-KI3thzKYPUwIKWobN-C:1567995115)*

**Comment by bholley** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/style/attr.rs, line 300 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3o-25zN569UinRp6j:-KI46rINDCgbquSkY2Wq:-499146009) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/components/style/attr.rs#L300)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> nit: Probably this is cleaner removing the `cfg` at the top and here: `impl ::std::ops::Deref`, leaving just one `#[cfg]`.
</details>
Done.

---

*[ports/geckolib/string_cache/Cargo.toml, line 1 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3qXdrlCI3ixSVzC_f:-KI46tHVxuZrd4Ov3yPK:160394146) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/Cargo.toml#L1)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> I guess that if we're hosting this in-tree, then https://github.com/servo/gecko-atom should disappear, right?
</details>
Yes

---

*[ports/geckolib/string_cache/lib.rs, line 118 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3qre1Rg2FbiWpfpOt:-KI470yrUrV9fbqn_JvI:-1876471234) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/lib.rs#L118)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> I was going to ask for something like this... Seems like a decent way to implement the utf-8 borrowing without leaks, if it's not too space-expensive... Can't think of a less-expensive way of doing it right now, though I arguably haven't thought too much about it.
</details>
Yeah, we'll need to measure.

---

*[ports/geckolib/string_cache/lib.rs, line 219 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3rSA4VvCkwHdnZ04f:-KI47H0FoaOtUc6Q_A46:-499146009) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/lib.rs#L219)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> nit: This would (probably?) work  as `&*string`, and we use that a lot more through the rest of the tree.
</details>
Done.

---

*[ports/geckolib/string_cache/lib.rs, line 226 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3rizUklCEXZImnbln:-KI47HSDJrXyPxYWHc2D:-499146009) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/lib.rs#L226)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> nit: ditto
</details>
Done.

---

*[ports/geckolib/string_cache/regen_atom_macro.py, line 19 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI1e4iCNQHlRH_q5YlP-r3-19:-KI47JMNE6EM8Ei58oFi:-486402025) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/regen_atom_macro.py#L19)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> After the IRC talk, I'm fine with landing this as-is, but (for the record) this would probably require at some point either taking in account other manglings like win32's, or using bindgen (that relies in clang for it).
</details>
Right

---

*[ports/geckolib/string_cache/regen_atom_macro.py, line 31 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11242#-KI3sAIWr9aTP9p6-pS7:-KI47ghGLbSgsE7bZzGr:-2072016818) ([raw file](https://github.com/servo/servo/blob/d1d851c1818b37e14de49d0ab4889d35af6bc2b3/ports/geckolib/string_cache/regen_atom_macro.py#L31)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> Writing this in a single `extern "C"` block would make the resulting code a lot more readable. Ideally using something like:
> 
> ```python
> "#[link_name=\"%s\"] pub static %s: *mut nsIAtom;\n" % (symbolify(atom[0]), atom[0])
> ```
> 
> This is generated code though, and it's clear what it does, so I won't insist on this a lot :)
</details>
I'm just going to leave it for now, sine we'll probably rewrite this once we make bindgen happen at compile-time and support win32.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11242)*

**Pull request merged**